### PR TITLE
kernel version should be 4.19+; recommend version 5.8 for cgroup v2

### PR DESCRIPTION
**Pull request by pacoxu**
For kernel long term support, see https://wiki.linuxfoundation.org/civilinfrastructureplatform/start and https://endoflife.date/linux
-  4.4 & 4.19 are selected as kernel Super Long Term Support (SLTS), and the Civil Infrastructure Platform will provide support until at least 2026. 
   -  **EDITED removed 4.4** as CIP is not in https://www.kernel.org/category/releases.html and maintained in a different place.
- For [cgroup v2](https://kubernetes.io/docs/concepts/architecture/cgroups/), Kubernetes recommends to use 5.8 and later, and in runc docs, the minimal version is 4.15 and 5.2+ is recommended. 

Other comments that may be related:

> In Kubernetes 1.31, cgroup v1 is moved to maintenance mode and 4.14 LTS EOF in Jan 2024 (linux, LTS) , besides, centos 7 is EOL in June 30, 2024. I chosen 4.15 as runc.
> 
> - for kernel minimal version, choosing 4.15 as runc + cgroup v2 https://github.com/opencontainers/runc/blob/3778ae603c706494fd1e2c2faf83b406e38d687d/docs/cgroup-v2.md?plain=1#L24
>   - kernel >= 4.15 with CONFIG_CGROUP_DEVICE and CONFIG_CGROUP_BPF is required.  From https://github.com/containerd/containerd/pull/3799#issuecomment-555740694. 
>   - cpu (since Linux 4.15)
> - cgroup v2: Kubernetes recommended kernel version https://kubernetes.io/docs/concepts/architecture/cgroups/
> 
> More details can be found in https://github.com/kubernetes/kubernetes/issues/116799.
> 
> The v1.31 KEP https://github.com/kubernetes/enhancements/issues/4569
> -  https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4569-cgroup-v1-maintenance-mode/README.md KEP-4569: Moving cgroup v1 support into maintenance mode 
> 
> Other minimal kernel version candidates
> - kernel 4.5 announced that cgroup v2 is not experimental anymore, as it supports io/pids/memory.
> - runc recommends 5.2+ as 5.2 supports freezer.

### CODE REVIEW

File: validators/kernel_validator.go
Comment (neolit123) [inline validators/kernel_validator.go:70]: should be:
```
var (
...
)
```

File: validators/kernel_validator.go
Comment (neolit123) [inline validators/kernel_validator.go:90]: same here. better to have a `var (...) ` block.

File: validators/kernel_validator_test.go
Comment (neolit123) [inline validators/kernel_validator_test.go:55]:
```suggestion
			name:    "5.7.3 no version regex matches",
```

File: validators/types_unix.go
Comment (neolit123) [inline validators/types_unix.go:36]: can be called `VersionsNote`

from a discussion on slack we had:
> the checks are regexp based so could be 4.4 == ok, 4.19 == ok, any other >= 4.19 version == ok, any other version == err.

so  in this note we can say that all those 4.x LTS versions are supported but the recommendation is to use 4.15 or 5.8+.

**Review by neolit123 — COMMENTED** [review]
we could show an error for any version that is not in that list of Versions and we don't need a RecommendedVersions or warnings IMO.
the error message could be x is not in the list of recommended versions. %s, where %s is a VersionsNote from the KernelSpec API.

File: validators/types_unix.go
Comment (pacoxu) [reply validators/types_unix.go]:
```
VersionsNote: "kernel version should be in maintenance, at least 4.4 or 4.19+, and to use cgroup v2, 5.8+ is recommended.",
```

update it as above..

File: validators/kernel_validator.go
Comment (pacoxu) [inline validators/kernel_validator.go:93]: The final error message is like`kernel release %s is unsupported, and kernel version should be in maintenance, at least 4.4 or 4.19+, and to use cgroup v2, 5.8+ is recommended.`

Comment (neolit123) [reply validators/kernel_validator.go:93]: i think we can make it
```suggestion
	return fmt.Errorf("kernel release %s is unsupported. %s", k.kernelRelease, kSpec.VersionsNote)
```

so just put `.` and after that is the note.

File: validators/types_unix.go
Comment (neolit123) [inline validators/types_unix.go:34]:
```suggestion
		VersionsNote: "Recommended LTS versions from the 4.x series are 4.4 and 4.19. Any 4.x version newer than 4.19 is also supported. For cgroups v2 support, the minimal version is 4.15 and the recommended version is 5.8+",
```

adding a more verbose message.

File: validators/types_windows.go
Comment (neolit123) [inline validators/types_windows.go:32]:
```suggestion
		VersionsNote: "The kernel version should be >= '10.0.14393'",
```

File: validators/types_unix.go
Comment (neolit123) [inline validators/types_unix.go:33]: please move the comment or remove it:
> // Requires 5.8+, or newer for cgroup v2.

i think there is a disagreement between the comments.
is 5.2+ or 5.8+ recommended for cgroup v2?

File: validators/kernel_validator_test.go
Comment (neolit123) [inline validators/kernel_validator_test.go:54]: can you please also add a test case for 4.6 or something else that is between the LTS versions and returns an error?

File: validators/types_unix.go
Comment (pacoxu) [reply validators/types_unix.go:33]: We may take kubernetes 5.8 here. 

```
// Requires 5.8+, or newer for cgroup v2.
```

File: validators/kernel_validator_test.go
Comment (pacoxu) [reply validators/kernel_validator_test.go:59]: Added.

File: validators/types_unix.go
Comment (neolit123) [reply validators/types_unix.go:33]: can we remove the extra comment on this line:
````
Versions:     []string{`^4\.4.*$`, `^4\.19.*$`, `^4\.[2-9][0-9].*$`, `^([5-9]|[1-9][0-9]+)\.([0-9]+)\.([0-9]+).*$`}, // Requires 5.8+, or newer for cgroup v2
```

**Review by neolit123 — COMMENTED** [review]
just one final nit 
https://github.com/kubernetes/system-validators/pull/37/files#r1691979088
and LGTM

**Comment by pacoxu** [comment]
Updated.

**Review by neolit123 — COMMENTED** [review]
/lgtm
/approve

should we create a release after this PR merges?

**Comment by neolit123** [comment]
/hold
drop the hold if needed.

**Comment by pacoxu** [comment]
> should we create a release after this PR merges?

Yes. 


> /hold
> drop the hold if needed.

We may wait for another week or two in case SIG-Node owners have some inputs.

**Comment by neolit123** [comment]
> > should we create a release after this PR merges?
> 
> Yes.
> 
> > /hold
> > drop the hold if needed.
> 
> We may wait for another week or two in case SIG-Node owners have some inputs.

you should probably give them a ping on slack, or they will miss this.

File: validators/types_unix.go
Comment (SergeyKanzhelev) [inline validators/types_unix.go:33]: for this note, do we have a check on cgroupv2 that the minimal version is satisfied?

cc: @harche

Comment (pacoxu) [reply validators/types_unix.go:33]: The current statement is from 
1. https://github.com/opencontainers/runc/blob/ad5b481dace5cda8ca7c659b7717a15517333198/docs/cgroup-v2.md?plain=1#L23-L24 
2.  https://kubernetes.io/docs/concepts/architecture/cgroups/#requirements: `Linux Kernel version is 5.8 or later`

If the runc one is not accurate, we may change to `the minimal version is 5.8`

File: validators/kernel_validator_test.go
Comment (SataQiu) [inline validators/kernel_validator_test.go:46]:
```suggestion
			name:    "4.4.14+ first version regex matches",
```

**Review by SataQiu — APPROVED** [review]

File: validators/types_unix.go
Comment (SergeyKanzhelev) [reply validators/types_unix.go:33]: I'm fine with the message. Just was wondering if we need to add a validaiton in kubelet on start when cgroupv2 are used.

Comment (pacoxu) [reply validators/types_unix.go:33]: there are many kernel version validations in kube. Mainly related to some FG or sysctl.

We may add a warning first for kubelet using cgoup v2.

File: validators/kernel_validator_test.go
Comment (pacoxu) [reply validators/kernel_validator_test.go]: Updated.

**Comment by pacoxu** [comment]
According to some discussions in https://github.com/endoflife-date/endoflife.date/pull/5608#issuecomment-2268591430, the CIP seems to be not in https://www.kernel.org/category/releases.html and maintained in a different place(gitlab) than `torvalds/linux`.  

I prefer to remove the 4.4.* here.

**Comment by neolit123** [comment]
> According to some discussions in [endoflife-date/endoflife.date#5608 (comment)](https://github.com/endoflife-date/endoflife.date/pull/5608#issuecomment-2268591430), the CIP seems to be not in https://www.kernel.org/category/releases.html and maintained in a different place(gitlab) than `torvalds/linux`.
> 
> I prefer to remove the 4.4.* here.

+1 to keep only official lts

**Comment by pacoxu** [comment]
Updated in commit: [update according to kernel official active kernel releases](https://github.com/kubernetes/system-validators/pull/37/commits/020ec91cd005724e9625d632dade5c4550a9fa6c).

File: validators/types_unix.go
Comment (neolit123) [inline validators/types_unix.go:31]:
```suggestion
		// 4.19 is an active kernel Long Term Support (LTS) release, tracked in https://www.kernel.org/category/releases.html.
```

File: validators/types_unix.go
Comment (neolit123) [inline validators/types_unix.go:33]:
```suggestion
		VersionsNote: "Recommended LTS version from the 4.x series is 4.19. Any 5.x or 6.x versions are also supported. For cgroups v2 support, the minimal version is 4.15 and the recommended version is 5.8+",
```

**Review by neolit123 — COMMENTED** [review]
please squash the commits to 1 as well.

**Review by neolit123 — COMMENTED** [review]
/lgtm

File: validators/types_unix.go
Comment (pacoxu) [reply validators/types_unix.go]: https://github.com/kubernetes/kubernetes/pull/126595

**Comment by pacoxu** [comment]
/unhold
@neolit123  Could you trigger a release after v1.31.0 is released this week.

**Pull request merged**
